pkg/udevevent/event.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2018 The OpenEBS Author
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package udevevent
+
+/*
+#cgo LDFLAGS: -ludev
+#include <libudev.h>
+#include <stdlib.h>
+*/
+import "C"
+import (
+	"syscall"
+	"unsafe"
+
+	"github.com/golang/glog"
+	"github.com/openebs/node-disk-manager/cmd/controller"
+	"github.com/openebs/node-disk-manager/pkg/udev"
+)
+
+//MonitorEvent start monitoring on udev source.
+func MonitorEvent() {
+	var udev *C.struct_udev
+	udev = C.udev_new()
+	defer C.udev_unref(udev)
+	if udev == nil {
+		glog.Fatal("Error while creating new udev object")
+	}
+	monitorDevices(udev)
+}
+
+//processDevice will process udev_device structure,
+//fetch the info about that disk and send event details.
+func processDevice(dev *C.struct_udev_device) {
+	if dev == nil {
+		return
+	}
+	if C.udev_device_get_devnode(dev) == nil {
+		return
+	}
+	idType := C.CString(udev.UDEV_DEVTYPE)
+	defer freeCharPtr(idType)
+	devType := C.GoString(C.udev_device_get_property_value(dev, idType))
+	action := C.GoString(C.udev_device_get_action(dev))
+	if devType == udev.UDEV_SYSTEM && (action == udev.UDEV_ACTION_ADD || action == udev.UDEV_ACTION_REMOVE) {
+		serial := C.GoString(C.udev_device_get_property_value(dev, C.CString(udev.UDEV_SERIAL)))
+		model := C.GoString(C.udev_device_get_property_value(dev, C.CString(udev.UDEV_MODEL)))
+		vendor := C.GoString(C.udev_device_get_property_value(dev, C.CString(udev.UDEV_VENDOR)))
+		wwn := C.GoString(C.udev_device_get_property_value(dev, C.CString(udev.UDEV_WWN)))
+		uuid := controller.GetnerateUid(wwn, model, serial, vendor)
+		major := C.GoString(C.udev_device_get_property_value(dev, C.CString(udev.UDEV_MAJOR)))
+		minor := C.GoString(C.udev_device_get_property_value(dev, C.CString(udev.UDEV_MINOR)))
+		syspath := "/sys/dev/block/" + major + ":" + minor
+		eventDetails := map[string]string{}
+		eventDetails[udev.UDEV_UUID] = uuid
+		eventDetails[udev.UDEV_SYSPATH] = syspath
+		eventDetails[udev.UDEV_ACTION] = action
+		glog.Info("Disk ", action, " event ", uuid)
```

Comment:
It appears that action and uuid are interchanged in this print statement

Reply:
changed.